File: tests/unit/focus-score.test.ts
import { describe, expect, it } from "vitest";
import {
  computeFocusScore,
  tierForScore,
  MINOR_PENALTY,
  SEVERE_PENALTY,
  ABANDONMENT_GRACE_SECONDS,
} from "@/lib/focus-score";

describe("tierForScore", () => {
  it("maps boundaries to the right tier", () => {
    expect(tierForScore(100).key).toBe("flow");
    expect(tierForScore(95).key).toBe("flow");
    expect(tierForScore(94).key).toBe("pristine");
    expect(tierForScore(85).key).toBe("pristine");
    expect(tierForScore(84).key).toBe("steady");
    expect(tierForScore(70).key).toBe("steady");
    expect(tierForScore(69).key).toBe("fragmented");
    expect(tierForScore(40).key).toBe("fragmented");
    expect(tierForScore(39).key).toBe("compromised");
    expect(tierForScore(0).key).toBe("compromised");
  });
});

describe("computeFocusScore", () => {
  it("awards a perfect clean session", () => {
    const r = computeFocusScore({ targetSeconds: 1800, focusSeconds: 1800, breaches: [] });
    expect(r.score).toBe(100);
    expect(r.tier.key).toBe("flow");
    expect(r.xp).toBe(Math.floor(100 * 30 * 1.5));
  });

  it("applies minor and severe breach penalties", () => {
    const r = computeFocusScore({
      targetSeconds: 600,
      focusSeconds: 600,
      breaches: [{ severity: "minor" }, { severity: "severe" }],
    });
    expect(r.penalty).toBe(MINOR_PENALTY + SEVERE_PENALTY);
    expect(r.score).toBe(100 - MINOR_PENALTY - SEVERE_PENALTY);
  });

  it("only penalises abandonment past the grace window", () => {
    const within = computeFocusScore({
      targetSeconds: 600,
      focusSeconds: 600,
      breaches: [],
      abandonmentSeconds: ABANDONMENT_GRACE_SECONDS,
    });
    expect(within.abandonmentPenalty).toBe(0);

    const past = computeFocusScore({
      targetSeconds: 600,
      focusSeconds: 600,
      breaches: [],
      abandonmentSeconds: ABANDONMENT_GRACE_SECONDS + 12,
    });
    expect(past.abandonmentPenalty).toBe(12);
    expect(past.score).toBe(88);
  });

  it("clamps focus time to the target and never goes negative", () => {
    const over = computeFocusScore({ targetSeconds: 300, focusSeconds: 9000, breaches: [] });
    expect(over.focusSecondsInt).toBe(300);
    expect(over.score).toBe(100);

    const wrecked = computeFocusScore({
      targetSeconds: 300,
      focusSeconds: 300,
      breaches: Array.from({ length: 5 }, () => ({ severity: "severe" as const })),
    });
    expect(wrecked.score).toBe(0);
    expect(wrecked.xp).toBe(0);
  });

  it("keeps ms precision internally and floors seconds at the DB boundary", () => {
    const r = computeFocusScore({ targetSeconds: 600, focusSeconds: 599.87, breaches: [] });
    expect(r.focusSecondsInt).toBe(599);
    expect(Number.isInteger(r.xp)).toBe(true);
  });

  it("gives zero XP for non-earning tiers", () => {
    const r = computeFocusScore({
      targetSeconds: 600,
      focusSeconds: 600,
      breaches: [{ severity: "severe" }],
    });
    expect(r.tier.key).toBe("fragmented");
    expect(r.xp).toBe(0);
  });
});

File: android/app/src/test/java/app/stackd/feature/room/session/FocusScoreTest.kt
package app.stackd.feature.room.session

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Parity checks against the web's `computeFocusScore` — a score computed here
 * and one computed on the web for the same session land in the same
 * `focus_history` table and the same leaderboards, so they must agree exactly.
 *
 * Values are hand-derived from the spec, not from the implementation, so a
 * regression in the port is caught rather than blessed.
 */
class FocusScoreTest {

    @Test
    fun `perfect session earns the flow tier and full xp`() {
        // 30 min held perfectly: 1800/1800*100 = 100 → flow (x1.5).
        // xp = floor(100 * (1800/60) * 1.5) = floor(4500).
        val r = FocusScore.compute(
            targetSeconds = 1800.0,
            focusSeconds = 1800.0,
            severeBreaches = 0,
            minorBreaches = 0,
        )
        assertEquals(100, r.score)
        assertEquals(FocusScore.Tier.FLOW, r.tier)
        assertEquals(4500, r.xp)
        assertEquals(FocusScore.SCORING_VERSION, r.scoringVersion)
    }

    @Test
    fun `half the target with no breach is fragmented and earns no xp`() {
        // 900/1800*100 = 50 → fragmented (x0). Multiplier zero ⇒ zero xp.
        val r = FocusScore.compute(1800.0, 900.0, 0, 0)
        assertEquals(50, r.score)
        assertEquals(FocusScore.Tier.FRAGMENTED, r.tier)
        assertEquals(0, r.xp)
    }

    @Test
    fun `tier reads the unrounded score, not the rounded one (v2)`() {
        // Construct a raw 84.5: focus/target*100 = 84.5, no penalty. It rounds
        // to 85, but v2 must classify it as STEADY (< 85), not PRISTINE — the
        // whole point of the version bump. xp uses the unrounded 84.5.
        // 845/1000*100 = 84.5, steady (x0.5).
        val r = FocusScore.compute(1000.0, 845.0, 0, 0)
        assertEquals(85, r.score) // display value rounds
        assertEquals(FocusScore.Tier.STEADY, r.tier) // reward reads 84.5
        // xp = floor(84.5 * (845/60) * 0.5) = floor(84.5 * 14.0833.. * 0.5)
        assertEquals(595, r.xp)
    }

    @Test
    fun `breaches subtract fixed penalties`() {
        // Full focus (100) minus one severe (40) and one minor (10) = 50.
        val r = FocusScore.compute(1800.0, 1800.0, severeBreaches = 1, minorBreaches = 1)
        assertEquals(50, r.score)
        assertEquals(50, r.penalty)
        assertEquals(FocusScore.Tier.FRAGMENTED, r.tier)
    }

    @Test
    fun `abandonment past the grace window subtracts second-for-second`() {
        // 40s abandonment, 15s grace ⇒ 25 penalty. 100 - 25 = 75 → steady.
        val r = FocusScore.compute(1800.0, 1800.0, 0, 0, abandonmentSeconds = 40.0)
        assertEquals(75, r.score)
        assertEquals(25, r.abandonmentPenalty)
        assertEquals(FocusScore.Tier.STEADY, r.tier)
    }

    @Test
    fun `abandonment within grace is free`() {
        val r = FocusScore.compute(1800.0, 1800.0, 0, 0, abandonmentSeconds = 15.0)
        assertEquals(0, r.abandonmentPenalty)
        assertEquals(100, r.score)
    }
}
